src/SessionSetup/ConnectToTargetDialog.cpp

```diff
@@ -39,9 +42,22 @@ ConnectToTargetDialog::ConnectToTargetDialog(
 ConnectToTargetDialog::~ConnectToTargetDialog() {}
 
 std::optional<TargetConfiguration> ConnectToTargetDialog::Exec() {
+  if (absl::GetFlag(FLAGS_launched_from_vsi)) {
+    connection_metric_ = std::make_unique<orbit_metrics_uploader::ScopedMetric>(
+        metrics_uploader_, orbit_metrics_uploader::OrbitLogEvent::ORBIT_CONNECT_TO_VSI_TARGET);
+  } else {
+    connection_metric_ = std::make_unique<orbit_metrics_uploader::ScopedMetric>(
+        metrics_uploader_, orbit_metrics_uploader::OrbitLogEvent::ORBIT_CONNECT_TO_CLI_TARGET);
+  }
```

Comment:
After looking at my other comments. If you want to get rid of the unique ptr, this can be done in the ctor like this:
```
connection_metric_(std::make_unique<orbit_metrics_uploader::ScopedMetric>(
          metrics_uploader_,
          (absl::GetFlag(FLAGS_launched_from_vsi)
               ? orbit_metrics_uploader::OrbitLogEvent::ORBIT_CONNECT_TO_VSI_TARGET
               : orbit_metrics_uploader::OrbitLogEvent::ORBIT_CONNECT_TO_CLI_TARGET)))
```

Reply:
See comment above: The main reason for the `unique_ptr` isn't initialization, but because it may be destructed elsewhere when an error occurs.

Looking at the code again, I get that this might be hard to understand. I could instead modify `OnAsyncDataAvailable` to return the error, as now all the error handling happens there.

Reply:
Actually, no, I can't, because `OnAsyncDataAvailable` is called as a result of a future...

Reply:
OnAsyncDataAvailable is allowed to have a return type.. your could do something along the lines 
```
joined_future.Then(main_thread_executor_.get(),
                     [this](MaybeSshAndInstanceData ssh_instance_data) {
                       auto result OnAsyncDataAvailable(std::move(ssh_instance_data));
                       if (result) {...}
                     });
```
or even
```
joined_future.Then(main_thread_executor_.get(),
                     [this](MaybeSshAndInstanceData ssh_instance_data) {
                       return OnAsyncDataAvailable(std::move(ssh_instance_data));
                     }).Then(main_thread_executor_.get(), (OnAsyncReturnType result) { ... });
```

But I think its not worth changing all the code, just to get rid of the unique_ptr.